library/manageiq_policy_assignment.py

```diff
@@ -0,0 +1,213 @@
+        entities_list = getattr(self.client.collections, entity_type)
+        return next((e.id for e in entities_list if e.name == entity_name), None)
+
+    def query_resource_policies_or_profiles(self, entity_type, resource_type, resource_id):
+        """ Returns the policies or policy profiles assigned to the resource.
+        """
+        try:
+            url = '{api_url}/{resource_type}/{resource_id}/{entity_type}?expand=resources'.format(api_url=self.api_url, resource_type=resource_type, resource_id=resource_id, entity_type=entity_type)
+            result = self.client.get(url)
+            return result.get('resources', [])
+        except Exception as e:
+            self.module.fail_json(msg="Failed to query resource {entity_type}: {error}".format(entity_type=entity_type, error=e))
+
+    def entity_assigned(self, entity_type, entity_id, resource_type, resource_id):
+        """Return True if the action is needed on the resource, False otherwise.
+        """
+        assigned_entities = self.query_resource_policies_or_profiles(entity_type, resource_type, resource_id)
+        return any(ae['id'] == entity_id for ae in assigned_entities)
+
+    def execute_action(self, entity_type, entity_id, resource_type, resource_id, action):
+        """Executes the action for the relevant entity on the resource.
+
+        Returns:
+            Whether or not a change took place and a message describing the
+            operation executed.
+        """
+        try:
+            href = '{api_url}/{entity_type}/{entity_id}'.format(api_url=self.api_url, entity_type=entity_type, entity_id=entity_id)
+            url = '{api_url}/{resource_type}/{resource_id}/{entity_type}'.format(api_url=self.api_url, resource_type=resource_type, resource_id=resource_id, entity_type=entity_type)
+            result = self.client.post(url, action=action, resource={'href': href})
+            if result['results'][0]['success']:
+                self.changed = True
+                return dict(
+                    changed=self.changed,
+                    msg=result['results'][0]['message']
+                )
+            else:
+                self.module.fail_json(msg="Failed to {action}: {fail_message}".format(action=action, entity=entity, fail_message=result['results'][0]['message']))
+        except Exception as e:
+            self.module.fail_json(msg="Failed to {action}: {error}".format(action=action, entity=entity, error=e))
+
+    def assign_or_unassign_entity(self, entity, entity_name, resource, resource_name, state):
+        """ Aassign or unassign the entity on the manageiq resource.
+
+        Returns:
+            Whether or not a change took place and a message describing the
+            operation executed.
+        """
+        entity_type = self.manageiq_entities[entity]
+        resource_type = self.manageiq_entities[resource]
+        entity_id = self.find_entity_by_name(entity_type, entity_name)
+        if entity_id:  # entity exists
```

Comment:
A pattern I like to follow is the "early exit" pattern.

Basically, exit the function as soon as possible instead of having complex if blocks like this. I think it makes the logic much easier to follow.

It will also keep pylint from saying "too many branches" since there are way fewer nested branches.

So on failure checks, have them run as the default case (normally this would be a return, but it's fine in a module to have it be module.fail_json).

So, in this specific case, it'd looks something like this (untested):

```python

    def assign_or_unassign_entity(self, entity, entity_name, resource, resource_name, state):
        """ Assign or unassign the entity on the manageiq resource.
        Returns:
            Whether or not a change took place and a message describing the
            operation executed.
        """
        entity_type = self.manageiq_entities[entity]
        resource_type = self.manageiq_entities[resource]
        entity_id = self.find_entity_by_name(entity_type, entity_name)

        if not entity_id:  # entity doesn't exist
            self.module.fail_json(
                msg="Failed to {action} {entity}: {entity_name} does not exsit in manageiq".format(action=ManageIQ.policy_actions[state], entity=entity, entity_name=entity_name))


        resource_id = self.find_entity_by_name(resource_type, resource_name)
        if not resource_id:  # resource doesn't exist
            self.module.fail_json(
                msg="Failed to {action} {entity}: {resource_name} {resource} does not exsit in manageiq".format(action=ManageIQ.policy_actions[state], entity=entity, resource_name=resource_name, resource=resource))


        assigned = self.entity_assigned(entity_type, entity_id, resource_type, resource_id)
        if assigned and state == 'absent':
            return self.execute_action(entity_type, entity_id, resource_type, resource_id, 'unassign')


         if (not assigned) and state == 'present':
            return self.execute_action(entity_type, entity_id, resource_type, resource_id, 'assign')


         # Default case is that there's nothing to change
         return dict(
                        changed=self.changed,
                        msg="{entity_name} {entity} already {action}ed".format(entity_name=entity_name, entity=entity, action=ManageIQ.policy_actions[state]))
```